---
title: "Pre_vs_Post_dea"
author: "Javier Moreno Venegas"
date: "25/4/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Introduction

This is a differential expression analysis of breast cancer tumour samples to compare gene expression patterns before and after de NAC.

The protocol is base on the execution of three R scripts that perform the following steps:
1. Load RCC files. Build counts matrix
2. Data normalisation
3. DE analysis

### Setting up R session  

```{r, message=FALSE, warning=FALSE, include=FALSE, paged.print=FALSE}
# set seed to be reproduciblish
set.seed(12345)

# set working directory
workingDir <- "."
setwd(workingDir)

options(stringsAsFactors = FALSE)

# Load required libraries
source(paste(getwd(), "Code/0_loadLibraries.R", sep='/'))
source("Code/coreDEA_1.R")
source("Code/coreDEA_2.R")
source("Code/coreDEA_3.R")
source("Code/coreDEA_4.R")
loadpkg("rgl")
loadpkg("calibrate") # To label the volcano plot
```  

### Execution
#### Load the data
Get a matrix with the raw counts from the set of samples we want to analyse. In this case we select pre NAC samples and their correspondent RCC files to build an RCCSet and get the counts matrix

```{r}
DataDir <- paste("Data", sep = '/') # ../../data
rccDir <- file.path(DataDir, "RCCData")

## Process file names: <sample>_<key1>_<key2>_<subtype>.RCC
## We have 204 files. There are 6 files without <subtype> : <sample>_<key1>_<key2>.RCC
## Key 1 contains pre/post info. 0 = PRE ; 1 = POST
metadata <-  data.frame(fname = list.files(path = rccDir, pattern="*.RCC")) %>%
  tidyr::separate(fname,c("Sample.name","key1","other"),sep="_",remove=F) %>%
  dplyr::select(fname, Sample.name, key1) %>%
  dplyr::filter(key1 == 0 | key1 == 1)

eset <- getRCCSet(metadata)
```  

We are working with `r nrow(metadata)` samples.

#### QC Control an normalisation.
Even though the samples we analyse here have passed the QC checks via nSolver. We monitor some QC aspects and normalise the data according to several criteria.

##### Imaging QC
Imaging QC refers to the percentage of fields of view (FOVs) successfully counted by a Digital Analyzer scan of a lane. When a substantial percentage of FOVs are not successfully counted, there may be issues with the resulting data (see nSolver User Manual).  

```{r echo=FALSE, message=F, warning=F}
plotFOV.key1(eset = eset, metadata = metadata, fov_threshold = 80)

# As there is a sample under the % FOV counted threshold we remove it
metadata <- pData(eset) %>% 
  tibble::rownames_to_column(var = "Sample.name") %>%
  inner_join(metadata, by="Sample.name") %>%
  dplyr::filter(FovCounted/FovCount * 100 > 80) %>%
  dplyr::select(fname, Sample.name, key1)
```  

##### Binding density QC
Due to the nature of the nCounter technology, analysis of some samples may produce too many or too few probes to be accurately counted by the Digital Analyzer. When too many probes are present, the Digital Analyzer is not able to distinguish each and every probe present in the lane. When too few fluorescent species are present, the Digital Analyzer may have difficulty focusing on the lane surface. Therefore, a measurement of mean binding density (spots per square micron) is provided with each lane scanned. The linear range of counting extends from 0.05 to 2.25 spots per square micron for assays run on an nCounter MAX or FLEX system. The range is 0.05 to 1.18 spots per square micron for assays run on the nCounter SPRINT system.  (see nSolver User Manual).

```{r message=FALSE, warning=FALSE, echo=FALSE}
plotBD.key1(eset = eset, metadata = metadata)
# All samples are in the binding density range
```  

##### Positive Controls

Check the expression of the positive controls. The positive genes follow the expeted pattern of expresssion.
```{r echo=FALSE, message=FALSE, warning=FALSE}
boxplot_expr(eset,is_positive)
```


Check the expression of the negative controls
```{r message=FALSE, warning=FALSE}
boxplot_expr(eset,is_negative)
```  

##### Noise Threshold  

We establish a noise threshold. This threshold is based on the mean and standard deviation of counts of the negative control genes and represents the background noise. We define it as the mean expression of the negative genes counts + 2 times the standard deviation.  

```{r echo=FALSE, comment=F, warning=F}
lodcounts = extract_pred(eset, is_negative)
lod = mean(lodcounts$count) + 2 * sd(lodcounts$count)
```  

##### Housekeeping genes  

Expression of each housekeeping genes in all samples. The line in red represents the noise threshold.

```{r echo=FALSE}
boxplot_expr(eset,is_housekeeping) + geom_hline(yintercept = (lod),colour="red")
```  
We can observe that all the housekeeping genes have enough expression (above the noise threshold).  

##### Expression of all the housekeeping genes in each sample.  

We plot the mean expression of all the housekeeping genes in each sample.  

```{r echo=FALSE, message=F, warning=F}
hK_mean_exprs_plot(eset, metadata, lod)
```  
We can obverve that some sample have overall low expression for the housekeeping genes, but all are above the noise threshold.  

##### General expression

Expression of all genes (Endogenous + Housekeeping).

```{r echo=FALSE, warning=F, message=F}
general_exprs_plot(eset, metadata, noise_threshold = lod)
```

#### Normalisation
##### Positive Normalization

We perform a normalization using the expression of the positive genes. This attempts to normalize for technical noise across the samples  

```{r echo=FALSE}
geo_pos = function(eset, metadata) {
     counts <-  exprs(eset)[, metadata$Sample.name]
     hk = counts[grepl("Positive", rownames(counts)), ]
     geoMeans = apply(hk, 2, function(col) exp(mean(log(col[col != 0]))))
     return(geoMeans)
 }
posFactor = function(eset, metadata) {
     geoMeans = geo_pos(eset, metadata)
     nf = mean(geoMeans)/geoMeans
     return(nf)
}
metadata$pos_nf = posFactor(eset, metadata)

counts = exprs(eset)[, metadata$Sample.name]

counts = counts[!grepl("Positive", rownames(counts)),]
counts = counts[!grepl("Negative", rownames(counts)),]

prenorm = counts %>% data.frame() %>% tidyr::gather("sample", "count")
prenorm$sample<-gsub("X","",prenorm$sample)
ggplot(prenorm, aes(sample, count)) + geom_boxplot(colour = "black", fill = "#56B4E9",outlier.size = 0.5) +
     scale_y_continuous(trans = "log2") + 
     xlab("") + ggtitle("pre-normalization") + theme(axis.text.x = element_text(angle = 90, 
     hjust = 1,vjust = 0.5,size = 5))+
    geom_hline(yintercept = lod, colour="red")
```


```{r echo=FALSE}
ncounts = counts %*% diag(metadata$pos_nf)
colnames(ncounts) = colnames(counts)
postnorm = ncounts %>%
  data.frame() %>%
  tidyr::gather("sample", "count")
postnorm$sample<-gsub("X","",postnorm$sample)
ggplot(postnorm, aes(sample, count)) +
  geom_boxplot(colour = "black", fill = "#56B4E9",outlier.size = 0.5) +
  scale_y_continuous(trans = "log2") +
  xlab("") + ggtitle("post-normalization") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1,vjust = 0.5,size = 5))+
  geom_hline(yintercept = lod,colour="red")
```  

##### Housekeeping normalization  

We select thouse housekeeping genes that have expression values greater than the noise threshold and a mean value of expression of at least 200 counts.

```{r echo=FALSE}
hk = ncounts[grepl("Housekeeping", rownames(ncounts)),]
abovenoise = rowSums(hk > (lod)) >= (ncol(hk))
hk_abovenoise = hk[abovenoise,]
aboveMean = (apply(hk_abovenoise,1,mean))>= 200
hk_sel= hk_abovenoise[aboveMean,]
hk_norm=rownames(hk_sel)
```

Out of the `r length(rownames(hk))` housekeeping genes, only `r length(rownames(hk_sel))` pass these criteria.

Using those genes, we observe that samples obtain more similar profiles.
```{r echo=FALSE, message= FALSE, warning= FALSE}
hk_pos = function(counts, hk_norm) {
  hk = counts[hk_norm,]
  geoMeans = apply(hk, 2, function(col) exp(mean(log(col[col != 0]))))
  return(geoMeans)}
hkFactor = function(counts, hk_norm) {
  geoMeans = hk_pos(counts, hk_norm)
  nf = mean(geoMeans) / geoMeans
  return(nf)}

metadata$hk_nf = hkFactor(ncounts, hk_norm)
ncounts = ncounts %*% diag(metadata$hk_nf)
colnames(ncounts) = colnames(counts)
postnorm = ncounts %>%
  data.frame() %>%
  tidyr::gather("sample", "count")
postnorm$sample<-gsub("X","",postnorm$sample)

ggplot(postnorm, aes(sample, count)) + geom_boxplot(colour = "black", fill = "#56B4E9",outlier.size = 0.5) +
     scale_y_continuous(trans = "log2") + 
     xlab("") + ggtitle("") + theme(axis.text.x = element_text(angle = 90, 
     hjust = 1,vjust = 0.5,size = 5))+
    geom_smooth(se=T, aes(group=1))+
    geom_hline(yintercept = lod,colour="red")
```  

##### Drop genes and produce the normalised counts matrix
Weâ€™ll drop the genes that are below the LOD in over 80% of the samples:

```{r echo=FALSE}
allNames <- rownames(ncounts)
ncounts <-  ncounts[(rowSums(ncounts < lod) < round((0.2 * ncol(ncounts)),0)),]
filteredNames <- rownames(ncounts)
filterOutGenes <- allNames[allNames%!in%filteredNames]
```

We save the metadata dataframe and the normalised counts matrix for further analysis.
```{r echo=FALSE}
save(metadata, file = "Data/RData/NAC_metadata.RData")
save(ncounts, file = "Data/RData/NAC_HKnormalised_counts.RData")
```  

##### PCA  

A PCA (Principal Component Analysis) performs a transformation over the data in order to obtain orthogonal vectors in such a way that the first principal component has the largest possible variance (that is, accounts for as much of the variability in the data as possible), and each succeeding component in turn has the highest variance possible under the constraint that it is orthogonal to the preceding components.

Plotting the samples in this transformed system show which samples are more similar.

Here we scaled and centered the normalized data and then performed PCA to look at how the samples cluster.

```{r echo=FALSE}
load(file = "Data/RData/NAC_metadata.RData")
load(file = "Data/RData/NAC_HKnormalised_counts.RData")

pca_loadings = function(object, ntop = 500) {
    object = as.matrix(object)
    rv <- matrixStats::rowVars(object)
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
    pca <- prcomp(t(object[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    names(percentVar) = colnames(pca$x)
    pca$percentVar = percentVar
    return(pca)
}
pcdata = scale(ncounts, center = TRUE, scale = TRUE)
pc = pca_loadings(pcdata, 50)
comps = data.frame(pc$x)
comps$Name = rownames(comps)

comps = comps %>% inner_join(metadata, by = c(Name = "Sample.name"))
pca_plot = function(comps, nc1, nc2, colorby) {
    c1str = paste0("PC", nc1)
    c2str = paste0("PC", nc2)
    ggplot(comps, aes_string(c1str, c2str, color = colorby)) + geom_point() + 
        theme_bw() + xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), 
        "% variance")) + ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 
        100), "% variance")) + theme(aspect.ratio=1)
}



## Variance explained by components
ggplot(data.frame(component=reorder(names(pc$percentVar), -pc$percentVar), 
                   percent_var=pc$percentVar),
                   aes(component, percent_var)) + 
   geom_bar(stat='identity') +
   ylab("percent of total variation") + 
   xlab("") + 
   theme_bw()

pca_plot(comps, 1, 2, "key1")
```


```{r echo=FALSE}
pca_plot(comps, 1, 3, "key1")
```


The following chunk makes a 3D PCA plot of the first three components. We can see that there is some clustering of the samples by pCR along the components 1 and 3.

```{r echo=TRUE, eval=F}
colorbykey1 <- function(k){
        if(k==0){return("red")}
        if(k==1){return("blue")}
}
comps$color = sapply(comps$key1, colorbykey1)
pca_plot3D = function(comps) {
        
    plot3d(comps$PC1, comps$PC2, comps$PC3, type="s",  size=0.75, lit=FALSE, xlab="", ylab="", zlab="", col = comps$color)
        mtext3d(paste0("PC1: ", round(pc$percentVar[1] * 100), "% variance"),       edge="x--", line=2)
        mtext3d(paste0("PC2: ", round(pc$percentVar[2] * 100), "% variance"), edge="y+-", line=3)
        mtext3d(paste0("PC3: ", round(pc$percentVar[3] * 100), "% variance"),          edge="z--", line=3)
    # + geom_point() + 
    #     theme_bw() + xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), 
    #     "% variance")) + ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 
    #     100), "% variance"))
}

pca_plot3D(comps=comps)
```  

#### DEA  
DEA by the DESeq2 library. The ncounts matrix is converted to integers for the analysis. The original counts are preserved in counts(dds). 

See [this thread](https://www.biostars.org/p/101727/) for log2FC.
```{r echo=FALSE, warning=F, message=F}
load(file = "Data/RData/NAC_metadata.RData")
load(file = "Data/RData/NAC_HKnormalised_counts.RData")

design = ~ key1
dea.o <- getDE(ncounts = round(ncounts), metadata = metadata, design = design)
#write_csv(dea.o, "Results/NAC-DEA.csv")
dea.p <- dea.o %>% dplyr::filter(padj <0.05) %>% arrange(-log2FoldChange)
View(dea.p)
```

We see the `r nrow(dea.p)` genes sorted by log2FC with significantly altered expression associated with NAC status. We can show the results of the differential expression analysis as a volcano plot. A volcano plot typically plots some measure of effect on the x-axis (typically the fold change) and the statistical significance on the y-axis (typically the -log10 of the p-value). Genes that are highly dysregulated are farther to the left and right sides, while highly significant changes appear higher on the plot.

```{r message=FALSE, warning=FALSE}
tab = data.frame(logFC = dea.o$log2FoldChange, negLogPval = -log10(dea.o$padj))
tab2 = data.frame(logFC = dea.o$log2FoldChange, negLogPval = -log10(dea.o$padj), Gene=dea.o$gene.name)
lfc = 1
pval = 0.05
par(mar = c(5, 4, 4, 4))
plot(tab, pch = 16, cex = 0.6, xlab = expression(log[2]~fold~change), ylab = expression(-log[10]~pvalue))

points(tab[(abs(tab$logFC) > lfc), ], pch = 16, cex = 0.8, col = "orange") 
points(tab[(tab$negLogPval > -log10(pval)), ], pch = 16, cex = 0.8, col = "green") 
points(tab[(abs(tab$logFC) > lfc & tab$negLogPval > -log10(pval)), ], pch = 16, cex = 0.8, col = "red") 
abline(h = -log10(pval), col = "green3", lty = 2) 
abline(v = c(-lfc, lfc), col = "blue", lty = 2) 
mtext(paste("pval =", pval), side = 4, at = -log10(pval), cex = 0.8, line = 0.5, las = 1) 
mtext(c(paste("-", lfc, "fold"), paste("+", lfc, "fold")), side = 3, at = c(-lfc, lfc), cex = 0.8, line = 0.5)

with(subset(tab2, negLogPval > -log10(pval) & abs(logFC)>1), textxy(logFC, negLogPval, labs=Gene, cex=.4))

# Save significant genes into a file
signGenes = (abs(tab$logFC) > lfc & tab$negLogPval > -log10(pval))
#write.csv(tab2[which(signGenes),], file = "Results/signGenes.csv")
```  

#### Pathway Analysis

Here we use GSEA to perform pathway analysis on the list of genes. Gene Set Enrichment Analysis focuses on the changes of expression in groups of genes, and by doing so, this method resolves the problem of the undetectable, small changes in the expression of single genes [see wikipedia.](https://en.wikipedia.org/wiki/Gene_set_enrichment_analysis). 


```{r echo=FALSE,message=F,warning=F}
load(file = "Data/RData/NAC_metadata.RData")
load(file = "Data/RData/NAC_HKnormalised_counts.RData")

design = ~ key1
res <- getDE.raw(ncounts = round(ncounts), metadata = metadata, design = design) # the core script for the pathway analysis expects a dataframe with the DEA named res
converted = unlist(lapply(strsplit(res$rowname, "_", fixed = TRUE), "[", 4))
converted = unlist(lapply(strsplit(converted, ".", fixed = TRUE), "[", 1))
converted = paste0("NM_", converted)
res$refseq_mrna = converted

lost.genes = length(res$rowname) - length((res %>% data.frame() %>% left_join(entrez, by = "refseq_mrna") %>% 
        filter(!is.na(entrezgene)) %>% filter(!is.na(log2FoldChange)) %>% filter(!is.na(lfcSE)))$rowname)

#gsea_rs = gsea_cp(res, "NACstatus")

#gsea_summary = gsea_rs$summary %>% arrange(pvalue)
#gsea_summary = convert_gsea_results(gsea_summary)
#write_csv(x = gsea_summary, path = "Results/NACstatus_gsea.csv")
```

We need to first pull out the Refseq IDs from the Nanostring IDs and convert those to Entrez IDs. Not all of the genes have Entrez IDs, however, we lose `r lost.genes` genes through this conversion. Genes are sorted by signal to noise ratio, which are defined as the log2Fold change (the signal) divided by the log2Fold change standard error (the noise).

I sorted the ontology/KEGG pathway terms by p-value and output a table of the results. The core_symbols are the genes that most strongly contributed to the pathway score, so the ones with the highest or lowest enrichment score depending on the direction of enrichment.

We perform a classic enrichment analysis in GO, Reactome and KEGG for all the genes differentially epressed by NAC.

```{r eval=FALSE, message=F, warning=F, include=FALSE, echo=TRUE}
enrich_rs = enrich_cp(res, "NAC", type="all")
enrich_summary = enrich_rs$summary %>% arrange(p.adjust)
enrich_summary = convert_enriched_ids(enrich_summary,entrezsymbol = entrezsymbol) %>% arrange(p.adjust)
write_csv(x = enrich_summary,path = "Results/NACstatus_enrichment.csv" )
save(enrich_rs, file = 'Data/RData/enrich_rs.RData')
```


We perform also a classic enrichment analysis for GO, Reactome and KEGG using the genes substantially over-expressed and under-expressed. We will use a two-fold fold-change cutoff.
```{r,eval=FALSE ,echo=TRUE,message=F, warning=F}
enrichover_rs = enrich_cp(res, "NAC", type="over")
enrichover_summary = enrichover_rs$summary %>% arrange(p.adjust)
enrichover_summary = convert_enriched_ids(enrichover_summary,entrezsymbol = entrezsymbol) %>% arrange(p.adjust)
write_csv(x = enrichover_summary,path = "Results/NACstatus_over_enrichment.csv" )
save(enrichover_rs, file = 'Data/RData/enrichover_rs.RData')

enrichunder_rs = enrich_cp(res, "NAC", type="under")
enrichunder_summary = enrichunder_rs$summary %>% arrange(p.adjust)
enrichunder_summary = convert_enriched_ids(enrichunder_summary,entrezsymbol = entrezsymbol) %>% arrange(p.adjust)
write_csv(x = enrichunder_summary,path = "Results/NACstatus_under_enrichment.csv" )
save(enrichunder_rs, file = 'Data/RData/enrichunder_rs.RData')
```

Finally we can see the GO BP, Reactome and KEGG pathways enriched.  

```{r}
load('Data/RData/enrich_rs.RData')
load('Data/RData/enrichover_rs.RData')
load('Data/RData/enrichunder_rs.RData')
```

GO BP Enrichment in all the significantly expressed genes:
```{r,echo=FALSE}
dotplot(enrich_rs$bp, x="count", showCategory=10, color="qvalue")
```

Reactome enrichment in all significantly expressed genes:
```{r,echo=FALSE}
dotplot(enrich_rs$re, x="count", showCategory=10, color="qvalue")
```

KEGG enrichment in all significantly expressed genes:
```{r,echo=FALSE}
dotplot(enrich_rs$kg, x="count", showCategory=10, color="qvalue")
```

GO BP Enrichment in over-expressed genes:
```{r,echo=FALSE}
dotplot(enrichover_rs$bp, x="count", showCategory=10, color="qvalue")
```

Reactome enrichment in over-expressed genes:
```{r,echo=FALSE}
dotplot(enrichover_rs$re, x="count", showCategory=10, color="qvalue")
```

KEGG enrichment in over-expressed genes:
```{r,echo=FALSE}
dotplot(enrichover_rs$kg, x="count", showCategory=10, color="qvalue")
```

GO BP Enrichment in under-expressed genes:
```{r,echo=FALSE}
dotplot(enrichunder_rs$bp, x="count", showCategory=10, color="qvalue")
```

Reactome enrichment in under-expressed genes:
```{r,echo=FALSE}
dotplot(enrichunder_rs$re, x="count", showCategory=10, color="qvalue")
```

KEGG enrichment in under-expressed genes:
```{r,echo=FALSE}
dotplot(enrichunder_rs$kg, x="count", showCategory=10, color="qvalue")
```


